File: container-with-most-water/container-with-most-water.go

package main

import (
  "fmt"
  "math"
)


func maxArea(height []int) int {
    max, left_index, right_index := 0, 0, len(height)-1
    for left_index < right_index {
        possible_max := int(math.Min(float64(height[left_index]), float64(height[right_index])) * float64(right_index-left_index))
        /*
        Open version of min and max functions
        possible_max := 0
        if height[left_index] < height[right_index] {
            possible_max = height[left_index] * (right_index - left_index)
            left_index += 1
        } else {
            possible_max = height[right_index] * (right_index - left_index)
            right_index -= 1
        }
        */
        if max < possible_max {
            max = possible_max
        }
        if height[left_index] < height[right_index] {
            left_index += 1
        } else {
            right_index -= 1
        }
    }
    return max
}

func main() {
    s := int[]{1,3,44,2,44}
    fmt.Println(maxArea(s))
}

File: container-with-most-water/container-with-most-water.java
import java.lang.*;

class Solution {
    public int maxArea(int[] height) {
        int maxArea , leftIndex, rightIndex;
        maxArea = leftIndex = 0;
        rightIndex = height.length-1;

        while(leftIndex < rightIndex){
            int newArea = Math.min(height[leftIndex], height[rightIndex]) * (rightIndex- leftIndex);
            maxArea = Math.max(newArea, maxArea);
            
            if(height[leftIndex] > height[rightIndex])
                --rightIndex;
            else
                ++leftIndex;
        }
        return maxArea; 
    
    }
}

class Main {
    public static void main(String[] args) {
        int [] height = {1,8,6,2,5,4,8,3,7};
        Solution sl = new Solution();
        sl.maxArea(height);
    }
}
